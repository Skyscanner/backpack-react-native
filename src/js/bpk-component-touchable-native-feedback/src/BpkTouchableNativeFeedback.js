/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016-2020 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow */

import React, { type Node, type ElementProps } from 'react';
import PropTypes from 'prop-types';
import {
  Platform,
  TouchableNativeFeedback,
  View,
  ViewPropTypes,
} from 'react-native';
import { colorSkyGrayTint02 } from 'bpk-tokens/tokens/base.react.native';

import { useBpkDynamicValue, type BpkDynamicValue } from '../../bpk-appearance';

type ViewProps = ElementProps<typeof View>;
type ViewStyleProp = $PropertyType<ViewProps, 'style'>;

export type Props = {
  children: Node,
  borderlessBackground: boolean,
  color: ?string | ?BpkDynamicValue<string>,
  style: ViewStyleProp,
};

const BpkTouchableNativeFeedback = (props: Props) => {
  const { children, style, borderlessBackground, color, ...rest } = props;
  const resolvedColor = useBpkDynamicValue(color);
  const defaultRippleColor = useBpkDynamicValue({
    light: null,
    dark: colorSkyGrayTint02,
  });
  const rippleColor = resolvedColor || defaultRippleColor;
  const preLollipop = Platform.Version < 21;
  const background = preLollipop
    ? TouchableNativeFeedback.SelectableBackground()
    : TouchableNativeFeedback.Ripple(rippleColor, borderlessBackground);

  return (
    // eslint-disable-next-line backpack/use-components
    <TouchableNativeFeedback style={style} background={background} {...rest}>
      {React.Children.only(children)}
    </TouchableNativeFeedback>
  );
};

BpkTouchableNativeFeedback.propTypes = {
  children: PropTypes.element.isRequired,
  borderlessBackground: PropTypes.bool,
  color: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.shape({
      light: PropTypes.string,
      dark: PropTypes.string,
    }),
  ]),
  style: ViewPropTypes.style,
};

BpkTouchableNativeFeedback.defaultProps = {
  borderlessBackground: true,
  color: null,
  style: null,
};

export default BpkTouchableNativeFeedback;
