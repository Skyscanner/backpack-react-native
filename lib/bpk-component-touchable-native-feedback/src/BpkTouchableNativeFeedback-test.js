/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016-2021 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow */

import React from 'react';
import renderer from 'react-test-renderer';
import { TouchableNativeFeedback } from 'react-native';

import { describeEachColorScheme } from '../../bpk-test-utils';

import BpkTouchableNativeFeedback from './BpkTouchableNativeFeedback';

const mockPlatform = (platform, version) => {
  const reactNative = jest.requireActual('react-native');
  jest
    .spyOn(reactNative.Platform, 'select')
    .mockImplementation((obj) => obj.platform || obj.default);
  reactNative.Platform.OS = platform;
  Object.defineProperty(reactNative.Platform, 'Version', {
    value: version,
  });
  return reactNative;
};

jest.mock(
  'react-native/Libraries/Components/Touchable/TouchableNativeFeedback',
  () =>
    jest.requireActual(
      'react-native/Libraries/Components/Touchable/TouchableNativeFeedback.js',
    ),
);

jest.unmock('../../bpk-appearance');

const FakeComponent = 'FakeComponent';

describe('BpkTouchableNativeFeedback', () => {
  beforeEach(() => {
    jest.resetModules();
    jest.spyOn(TouchableNativeFeedback, 'Ripple');
  });

  afterEach(() => {
    TouchableNativeFeedback.Ripple.mockReset();
  });

  const styles = () => {
    const { StyleSheet } = jest.requireActual('react-native');
    return StyleSheet.create({
      custom: {
        width: '300em',
      },
    });
  };

  describeEachColorScheme(
    BpkTouchableNativeFeedback,
    (WithColorScheme, colorScheme) => {
      it('should render correctly on API < 21', () => {
        mockPlatform('android', 19);

        const tree = renderer
          .create(
            <WithColorScheme>
              <FakeComponent />
            </WithColorScheme>,
          )
          .toJSON();

        expect(tree).toMatchSnapshot();
      });

      it('should render correctly on API >= 21', () => {
        mockPlatform('android', 21);

        const tree = renderer
          .create(
            <WithColorScheme>
              <FakeComponent />
            </WithColorScheme>,
          )
          .toJSON();

        expect(tree).toMatchSnapshot();
      });

      it('should render correctly with custom style prop', () => {
        mockPlatform('android', 21);
        const tree = renderer
          .create(
            <WithColorScheme style={styles().custom}>
              <FakeComponent />
            </WithColorScheme>,
          )
          .toJSON();

        expect(tree).toMatchSnapshot();
      });

      it('should render correctly with arbitrary props', () => {
        mockPlatform('android', 21);
        const tree = renderer
          .create(
            <WithColorScheme testID="arbitrary value">
              <FakeComponent />
            </WithColorScheme>,
          )
          .toJSON();

        expect(tree).toMatchSnapshot();
      });

      it('should render correctly with a string `color` prop', () => {
        mockPlatform('android', 21);
        const tree = renderer
          .create(
            <WithColorScheme color="rgb(255, 255, 255)">
              <FakeComponent />
            </WithColorScheme>,
          )
          .toJSON();

        expect(TouchableNativeFeedback.Ripple).toHaveBeenCalledWith(
          'rgb(255, 255, 255)',
          true,
        );
        expect(tree).toMatchSnapshot();
      });

      it('should render correctly with a dynamic `color` prop', () => {
        mockPlatform('android', 21);
        const tree = renderer
          .create(
            <WithColorScheme
              color={{ light: 'rgb(255, 255, 255)', dark: 'rgb(0, 0, 0)' }}
            >
              <FakeComponent />
            </WithColorScheme>,
          )
          .toJSON();

        expect(TouchableNativeFeedback.Ripple).toHaveBeenLastCalledWith(
          { light: 'rgb(255, 255, 255)', dark: 'rgb(0, 0, 0)' }[colorScheme],
          true,
        );
        expect(tree).toMatchSnapshot();
      });
    },
  );
});
