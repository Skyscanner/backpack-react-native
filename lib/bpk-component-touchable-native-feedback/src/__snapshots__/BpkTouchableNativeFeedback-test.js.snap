// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BpkTouchableNativeFeedback dark mode should render correctly on API < 21 1`] = `
<FakeComponent
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
/>
`;

exports[`BpkTouchableNativeFeedback dark mode should render correctly on API >= 21 1`] = `
<FakeComponent
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
/>
`;

exports[`BpkTouchableNativeFeedback dark mode should render correctly with a dynamic \`color\` prop 1`] = `
<FakeComponent
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
/>
`;

exports[`BpkTouchableNativeFeedback dark mode should render correctly with a string \`color\` prop 1`] = `
<FakeComponent
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
/>
`;

exports[`BpkTouchableNativeFeedback dark mode should render correctly with arbitrary props 1`] = `
<FakeComponent
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="arbitrary value"
/>
`;

exports[`BpkTouchableNativeFeedback dark mode should render correctly with custom style prop 1`] = `
<FakeComponent
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
/>
`;

exports[`BpkTouchableNativeFeedback light mode should render correctly on API < 21 1`] = `
<FakeComponent
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
/>
`;

exports[`BpkTouchableNativeFeedback light mode should render correctly on API >= 21 1`] = `
<FakeComponent
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
/>
`;

exports[`BpkTouchableNativeFeedback light mode should render correctly with a dynamic \`color\` prop 1`] = `
<FakeComponent
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
/>
`;

exports[`BpkTouchableNativeFeedback light mode should render correctly with a string \`color\` prop 1`] = `
<FakeComponent
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
/>
`;

exports[`BpkTouchableNativeFeedback light mode should render correctly with arbitrary props 1`] = `
<FakeComponent
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="arbitrary value"
/>
`;

exports[`BpkTouchableNativeFeedback light mode should render correctly with custom style prop 1`] = `
<FakeComponent
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
/>
`;
